overlord/snapstate/target.go

```diff
@@ -312,11 +312,103 @@ func (s *storeInstallGoal) toInstall(ctx context.Context, st *state.State, opts
 		})
 	}
 
-	// TODO: check validation sets here to catch invalid component revisions
+	for _, t := range installs {
+		sn, ok := s.find(t.info.InstanceName())
+		if !ok {
+			return nil, fmt.Errorf("internal error: snap to install was not requested: %s", t.info.InstanceName())
+		}
+
+		if err := checkTargetAgainstValidationSets(t, "install", sn.RevOpts.ValidationSets); err != nil {
+			return nil, err
+		}
+	}
 
 	return installs, err
 }
 
+func checkTargetAgainstValidationSets(target target, action string, vsets *snapasserts.ValidationSets) error {
+	constraints, err := vsets.Presence(target.info)
+	if err != nil {
+		return err
+	}
+
+	if err := checkSnapAgainstConstraints(target.info.InstanceName(), target.info.Revision, constraints, action); err != nil {
+		return err
+	}
+
+	comps := make(map[string]snap.Revision, len(target.components))
+	for _, comp := range target.components {
+		comps[comp.ComponentName()] = comp.Revision()
+	}
+
+	return checkComponentsAgainstConstraints(target.info.SnapName(), comps, constraints, action)
+}
+
+// should something like this maybe be a method on snapasserts.ValidationSets?
+// the error strings are the main thing that don't really fit well in there
```

Comment:
It seems fine as a standalone function, a method doesn't feel particularly more natural than this